### PR TITLE
Adding psnrb

### PR DESCRIPTION
**Pull request by soma2000-lang**
## What does this PR do?

Part of #799 

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by soma2000-lang** [comment]
@Borda @SkafteNicki  Any help regarding the failing tests would be really helpful

File: docs/source/image/peak_signl_to_noise [hunk 1]
```diff
@@ -0,0 +1,22 @@
+.. customcarditem::
+   :header: Peak Signal To Noise Ratio With Blocked Effectc
+   :image: https://pl-flash-data.s3.amazonaws.com/assets/thumbnails/image_classification.svg
+   :tags: Image
+
+.. include:: ../links.rst
+
+#############################
+Peak Signal To Noise Ratio With Blocked Effect
+#############################
```
Comment (SkafteNicki) [inline docs/source/image/peak_signl_to_noise:10]:
```suggestion
##############################################
Peak Signal To Noise Ratio With Blocked Effect
##############################################
```

File: docs/source/links.rst [hunk 2]
```diff
@@ -116,3 +116,4 @@
 .. _Signal-to-noise ratio: https://arxiv.org/abs/1811.02508
 .. _Permutation invariant training: https://arxiv.org/abs/1607.00325
 .. _Multilabel coverage error: https://link.springer.com/chapter/10.1007/978-0-387-09823-4_34
+.. _Peak Signal to Noise Ratio with Blocked Effect:https://ieeexplore.ieee.org/abstract/document/5535179
```
Comment (SkafteNicki) [inline docs/source/links.rst:119]: missing space
```suggestion
.. _Peak Signal to Noise Ratio with Blocked Effect: https://ieeexplore.ieee.org/abstract/document/5535179
```

File: src/torchmetrics/functional/image/psnrb.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    if not dim_list:
+        n_obs = tensor(target.numel(), device=target.device)
+    else:
+        n_obs = tensor(target.size(), device=target.device)[dim_list].prod()
+        n_obs = n_obs.expand_as(sum_squared_error)
+
+    return sum_squared_error, n_obs
+
+
+def peak_signal_noise_ratio(
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/image/psnrb.py:162]:
```suggestion
def peak_signal_noise_ratio_with_blocked_effect(
```

File: src/torchmetrics/image/psnrb.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+import torch
+from torch import Tensor, tensor
+from typing_extensions import Literal
+
+from torchmetrics.functional.image.psnr import _psnr_compute, _psnr_update
+from torchmetrics.metric import Metric
+from torchmetrics.utilities import rank_zero_warn
+
+
+class PeakSignalNoiseRatio(Metric):
```
Comment (SkafteNicki) [inline src/torchmetrics/image/psnrb.py:25]:
```suggestion
class PeakSignalToNoiseRatioWithBlockedEffect(Metric):
```

File: src/torchmetrics/functional/image/psnrb.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    elif dim == 2:
+        height, width = target.Size
+        channels = 1
+    else:
+        raise ValueError("Not a 1-channel/3-channel grayscale image")
+
+    if channels > 1:
+        raise ValueError("Not for color images")
+
+    h = torch.tensor(range(0, width - 1))
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/image/psnrb.py:38]:
```suggestion
    h = torch.arange(width - 1)
```

File: src/torchmetrics/functional/image/psnrb.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    h_b = torch.tensor(range(block_size - 1, width - 1, block_size))
+    h_bc = torch.tensor(list(set(h).symmetric_difference(h_b)))
+
+    v = torch.tensor(range(0, height - 1))
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/image/psnrb.py:42]:
```suggestion
    v = torch.arange(height - 1)
```

File: src/torchmetrics/functional/image/psnrb.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+
+    # D code
+    d_b /= n_hb + n_vb
+    d_bc /= n_hbc + n_vbc
+
+    # Log
+    if d_b > d_bc:
+        t = torch.log2(block_size) / torch.log2(min(height, width))
+    else:
+        t = 0
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/image/psnrb.py:83]:
```suggestion
    t = torch.log2(block_size) / torch.log2(min(height, width)) if d_b > d_bc else 0
```

File: src/torchmetrics/functional/image/psnrb.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+    # v_b for loop
+    for j in list(v_b):
+        diff = target[j, :] - target[j + 1, :]
+        d_b += torch.sum(torch.square(diff))
+
+    # V_bc for loop
+    for j in list(v_bc):
+        diff = target[j, :] - target[j + 1, :]
+        d_bc += torch.sum(tensor.square(diff))
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/image/psnrb.py:67]: I think this should be able to be vectorized in some way, possible using `torch.gather`

**Review by SkafteNicki — COMMENTED** [review]
Some comments, already looking good.
Please also add entry in changelog :)

File: src/torchmetrics/functional/image/psnrb.py (hunk 8 above)
Comment (soma2000-lang) [reply src/torchmetrics/functional/image/psnrb.py:67]: ok willl try

**Comment by soma2000-lang** [comment]
image

@SkafteNicki  I have almost made all the changes as you suggested ,but  I require some help as to what might have caused  this error while running pytest

**Comment by soma2000-lang** [comment]
@Borda @SkafteNicki  I cannot understand why its still giving the import error,even after the init file has been fixed.

**Comment by Borda** [comment]
> @Borda @SkafteNicki I cannot understand why its still giving the import error,even after the init file has been fixed.

seems you used the previous PSNR but did not update the class name, also curious if we could rather reuse some functionality instead of writing all from the ground... :rabbit:

**Comment by Borda** [comment]
also it seem you have not linked reference metrics as of:
```
tests/unittests/image/test_psnrb.py:94: [F821] undefined name '_skimage_psnrb'
tests/unittests/image/test_psnrb.py:95: [F821] undefined name '_base_e_sk_psnrb'
```

File: src/torchmetrics/functional/image/psnrb.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+import torch
+from torch import Tensor, tensor
+from typing_extensions import Literal
+
+from torchmetrics.utilities import rank_zero_warn, reduce
+
+
+def _compute_bef(
+    self, target: Tensor, dim: Optional[Union[int, Tuple[int, ...]]] = None, block_size=8
```
Comment (Borda) [inline src/torchmetrics/functional/image/psnrb.py:24]: Why self? also, is prediction missing?

**Review by Borda — CHANGES_REQUESTED** [review]

**Comment by soma2000-lang** [comment]
@Borda @SkafteNicki any other changes to be made??

**Review by Borda — APPROVED** [review]

File: src/torchmetrics/functional/image/psnrb.py [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+
+
+def _compute_bef(target: Tensor, dim: Optional[Union[int, Tuple[int, ...]]] = None, block_size=8) -> Tensor:
+    if dim == 3:
+        height, width, channels = target.Size
+    elif dim == 2:
+        height, width = target.Size
+        channels = 1
+    else:
+        raise ValueError("Not a 1-channel/3-channel grayscale image")
```
Comment (stancld) [inline src/torchmetrics/functional/image/psnrb.py:30]: In update method, we define `dim` as:
```
dim: Dimensions to reduce PSNR scores over provided as either an integer or a list of integers.
```
while here is `dim` as an identifier for target image shape, which is likely to cause issues.

File: src/torchmetrics/functional/image/psnrb.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+import torch
+from torch import Tensor, tensor
+
+
+def _compute_bef(x: Tensor, block_size: int = 8) -> Tensor:
+    """Compute block effect.
+
+    Args:
+        x: input image
+        dim: dimension to calculate block over
```
Comment (stancld) [inline src/torchmetrics/functional/image/psnrb.py:26]:
```suggestion
```

File: src/torchmetrics/functional/image/psnrb.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+    """
+    (
+        _,
+        channels,
+        height,
+        width,
+    ) = x.shape
+    if channels > 1:
+        raise ValueError(f"``psnrb metric expects grayscale images, but got an images with {channels} channels.")
```
Comment (stancld) [inline src/torchmetrics/functional/image/psnrb.py:44]:
```suggestion
        raise ValueError(f"`psnrb` metric expects grayscale images, but got an images with {channels} channels.")
```

File: src/torchmetrics/functional/image/psnrb.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def _psnrb_update(preds: Tensor, target: Tensor, block_size: int = 8) -> Tuple[Tensor, Tensor, Tensor]:
+    """Updates and returns variables required to compute peak signal-to-noise ratio.
+
+    Args:
+        preds: Predicted tensor
+        target: Ground truth tensor
+        block_size:
+            Integer indication the block size
```
Comment (stancld) [inline src/torchmetrics/functional/image/psnrb.py:96]:
```suggestion
        block_size: Integer indication the block size
```

**Review by stancld — APPROVED** [review]

File: tests/unittests/image/test_psnrb.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+            preds,
+            target,
+            PeakSignalNoiseRatioWithBlockedEffect,
+            peak_signal_noise_ratio_with_blocked_effect,
+        )
+
+
+def test_error_on_color_images():
+    """Test that appropriate error is raised when color images are passed to PSNRB metric."""
+    with pytest.raises(ValueError, match="``psnrb metric expects grayscale images.*"):
```
Comment (stancld) [inline tests/unittests/image/test_psnrb.py:98]:
```suggestion
    with pytest.raises(ValueError, match="`psnrb` metric expects grayscale images.*"):
```
